src/lib/device/hs_super_blk.h

```diff
@@ -0,0 +1,189 @@
+    }
+
+    std::string to_string() const {
+        auto str = fmt::format("magic={:#x}, checksum={}, first_blk_header=[{}], this_pdev_info=[{}]", get_magic(),
+                               checksum, hdr.to_string(), this_pdev_hdr.to_string());
+        return str;
+    }
+};
+#pragma pack()
+static_assert(sizeof(first_block) <= first_block::s_atomic_fb_size);
+
+/////////////// Overarching super block information ////////////////
+class hs_super_blk {
+public:
+    // Minium chunk size we can create in data device. Keeping this lower will increase number of chunks and thus
+    // area for super block will be higher.
+    static constexpr uint64_t MIN_CHUNK_SIZE_DATA_DEVICE = 16 * 1024 * 1024;
+
+    // Higher min chunk size than data device to ensure to limit max chunks in fast pdevs and thus lesser super block
+    // area on more expensive fast device.
+    static constexpr uint64_t MIN_CHUNK_SIZE_FAST_DEVICE = 32 * 1024 * 1024;
+
+    // Maximum number of chunks across all devices. We need to keep in mind the BlkId restriction (to address the
+    // chunks)
+    static constexpr uint32_t MAX_CHUNKS_IN_SYSTEM = 65536;
+
+    // Maximum vdevs in the system. Increasing this will have more vdev information in super block
+    static constexpr uint32_t MAX_VDEVS_IN_SYSTEM = 1024;
+
+    static constexpr uint64_t EXTRA_SB_SIZE_FOR_DATA_DEVICE = 8 * 1024 * 1024;
+    static constexpr uint64_t EXTRA_SB_SIZE_FOR_FAST_DEVICE = 1 * 1024 * 1024;
+
+    static constexpr uint32_t first_block_offset() { return 0; } // Offset in physical device we can use for first block
+    static uint32_t first_block_size() { return first_block::s_io_fb_size; }
+    static uint64_t vdev_super_block_size();
+    static uint64_t chunk_super_block_size(const dev_info& dinfo);
+    static uint64_t chunk_info_bitmap_size(const dev_info& dinfo) {
+        // Chunk bitmap area has bitmap of max_chunks rounded off to 4k page
+        return sisl::round_up(std::max(1u, hs_super_blk::max_chunks_in_pdev(dinfo) / 8), 4096);
+    }
+
+    static uint64_t total_size(const dev_info& dinfo) { return total_used_size(dinfo) + future_padding_size(dinfo); }
+    static uint64_t total_used_size(const dev_info& dinfo) {
+        return first_block_size() + vdev_super_block_size() + chunk_super_block_size(dinfo);
+    }
+    static uint64_t vdev_sb_offset() { return first_block_offset() + first_block_size(); }
+    static uint64_t chunk_sb_offset() { return vdev_sb_offset() + vdev_super_block_size(); }
+
+    static uint64_t future_padding_size(const dev_info& dinfo) {
+        return (dinfo.dev_type == HSDevType::Fast) ? EXTRA_SB_SIZE_FOR_FAST_DEVICE : EXTRA_SB_SIZE_FOR_DATA_DEVICE;
+    }
+    static uint32_t max_chunks_in_pdev(const dev_info& dinfo) {
+        return (dinfo.dev_size - 1) /
+            ((dinfo.dev_type == HSDevType::Fast) ? MIN_CHUNK_SIZE_DATA_DEVICE : MIN_CHUNK_SIZE_FAST_DEVICE) +
```

Comment:
I think it should:
(dinfo.dev_type == HSDevType::Fast) ?  MIN_CHUNK_SIZE_FAST_DEVICE : MIN_CHUNK_SIZE_DATA_DEVICE

Reply:
Good catch